### DISCUSSION TITLE
Unable to Resolve ITMS-90078 After Adding APNs Certificate: Push Notification Entitlement Issue in Codemagic Build

### DISCUSSION THREAD

**Discussion by gusentanan** [codemagic-ci-cd/codemagic-docs#2758] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

## Step to reproduce
Start the new build in codemagic

## Step I've taken
1. Generated and added APNs certificates for production and development in the Apple Developer portal.
2. Updated the provisioning profiles to include the new certificates and enabled the Push Notifications entitlement.
3. Ensured that the Push Notification capability is enabled and the aps-environment key is present in the `Runner.entitlements` file.
4. Set up Codemagic with the updated profiles and certificates, and confirmed automatic code signing.
5. Start build in codemagic

## Email received from Apple Store Connect
Hello,

We noticed one or more issues with a recent delivery for the following app.

ITMS-90078: Missing potentially required entitlement - Your app, or a library that’s included in your app, uses Apple Push Notification service (APNs) registration APIs, but the APS Environment Entitlement isn’t included in the app signature’s entitlements. If your app uses APNs, make sure the App ID is enabled for push notifications in Certificates, Identifiers & Profiles, sign the app with a distribution provisioning profile that includes the APS Environment Entitlement, and upload a new build to ensure that push notifications function as intended. For details, visit: https://developer.apple.com/documentation/usernotifications/registering-your-app-with-apns.

Apple Developer Relations

### Expected results

After completing the steps above, the app should be successfully uploaded to App Store Connect without triggering the ITMS-90078 error. The push notification entitlements should be correctly embedded in the app, allowing it to handle push notifications.

### Actual results

Despite following the steps, the upload to App Store Connect still fails with the ITMS-90078 error

### Build id (optional)

66f673c2fc9cb3d3d36d0e0d

**Comment by dtrdic** [comment]
hello @gusentanan, yes, based on the warning message you provided, it seems that your provisioning profile might not have the correct entitlement enabled for your bundle identifier. Can you double-check on the [apple developer page](https://developer.apple.com/account/resources/profiles/list) to see if the provisioning profile you’re referencing is marked as invalid?

One other thing that you can try.. there is a Codemagic CLI command, which allows you to add capabilities to your provisioning profile without visiting the Apple Developer page. You can include this in your workflow or run it locally
Please refer to [this documentation](https://github.com/codemagic-ci-cd/cli-tools/blob/master/docs/app-store-connect/bundle-ids/enable-capabilities.md#enable-capabilities)

Here's an example of how this step would look if you add it to your workflow pipeline:
```
- name: Install capabilities script: |
app-store-connect bundle-ids enable-capabilities \
B5J6V3KQ94 \
--capability "Push Notifications"
```

Just keep in mind that you need to use the **BUNDLE_ID_RESOURCE_ID**, not the actual identifier, and you can find it in the URL after opening the actual ID in the Apple Developer portal:
​https://developer.apple.com/account/resources/identifiers/bundleId/edit/ **BUNDLE_ID_RESOURCE_ID**

**Reply by gusentanan** [reply]
Hi, thanks for the response! I added the scripts to my workflow pipeline, but it seems like adding those scripts is causing the provisioning profile to become invalid each time I deploy to the App Store.

**Comment by jimsakiko** [comment]
Hi @gusentanan, I am in the same boat and exactly did your steps, verifying that the Provisioning Profile has the correct (APN) entitlement. But no joy. Have you been able to resolve this?

**Comment by gusentanan** [comment, marked as answer]
Hi, sorry for the delayed response, @jimsakiko. I was able to resolve the email warning by enabling the Push Notification capabilities through Xcode, which requires a MacBook. It seems that manually adding the `Runner.entitlements` file is not working. What worked for me was enabling the Push Notification capability in Xcode by following [this step](https://stackoverflow.com/a/63572813/19672390).

Once you've done that, Xcode will automatically update your `project.pbxproj` file under the `Runner.xcodeproj` folder and generate the correct `Runner.entitlements`. Afterward, just commit your changes and run the build in Codemagic.

P.S. Make sure you've also generated both development and production Push Notification certificates through the Apple Developer Portal, and register those in your messaging service (e.g., Firebase Messaging). Alternatively, you can use an auth key from the Apple Developer Portal as well.

**Reply by jimsakiko** [reply]
Thank you. I will resolve this with that guidance!

**Reply by serppower** [reply]
Is there any other way than using Xcode? what is the point from using codemagic then If I still need xcode? I am on a windows machine with android studio & flutter project.
